File: translation-library/nodejs/src/translation-library/chunker.ts
/**
 * Модуль для разбивки текста на чанки
 * Основан на реализации из SillyTavern
 */

import { TranslationError } from './types';

/**
 * Разбивает текст на чанки заданного размера
 * @param text Текст для разбивки
 * @param maxSize Максимальный размер чанка в символах
 * @returns Массив чанков
 */
export function chunkText(text: string, maxSize: number): string[] {
  if (!text || text.length <= maxSize) {
    return [text];
  }

  const chunks: string[] = [];
  let position = 0;

  while (position < text.length) {
    // Пытаемся разбить по предложению или абзацу
    let chunkEnd = Math.min(position + maxSize, text.length);

    // Если это не конец текста, пытаемся найти границу предложения
    if (chunkEnd < text.length) {
      // Ищем точку, восклицательный или вопросительный знак
      const sentenceEnd = text.slice(position, chunkEnd).search(/[.!?]\s/);
      if (sentenceEnd > maxSize * 0.5) {
        chunkEnd = position + sentenceEnd + 1;
      } else {
        // Ищем перенос строки
        const newlineEnd = text.slice(position, chunkEnd).search(/\n\s*/);
        if (newlineEnd > maxSize * 0.5) {
          chunkEnd = position + newlineEnd;
        }
      }
    }

    const chunk = text.slice(position, chunkEnd).trim();
    if (chunk) {
      chunks.push(chunk);
    }

    position = chunkEnd;
  }

  return chunks.length > 0 ? chunks : [text];
}

/**
 * Разбивает текст по предложениям
 * @param text Текст для разбивки
 * @returns Массив предложений
 */
export function splitBySentences(text: string): string[] {
  // Регулярное выражение для разделения по предложениям
  const sentenceRegex = /([^.!?]+[.!?]+["']?|\S+)/g;
  const matches = text.match(sentenceRegex) || [];
  
  return matches.map(s => s.trim()).filter(s => s.length > 0);
}

/**
 * Объединяет чанки обратно в текст
 * @param chunks Массив чанков
 * @returns Объединённый текст
 */
export function mergeChunks(chunks: string[]): string {
  return chunks.join(' ');
}

/**
 * Сохраняет ссылки на изображения при чанковании
 * @param text Текст для чанкования
 * @param maxSize Максимальный размер чанка
 * @returns Объект с чанками и ссылками
 */
export function chunkWithLinks(text: string, maxSize: number): {
  chunks: string[];
  links: string[];
} {
  // Разделяем текст по ссылкам на изображения
  const imageRegex = /!\[.*?\]\([^)]*\)/g;
  const chunks = text.split(imageRegex);
  const links = [...text.matchAll(imageRegex)].map(m => m[0]);

  const resultChunks: string[] = [];
  const resultLinks: string[] = [];

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    if (!chunk || chunk.trim().length === 0) {
      continue;
    }

    const chunkChunks = chunkText(chunk, maxSize);
    resultChunks.push(...chunkChunks);

    if (i < links.length) {
      resultLinks.push(links[i]);
    }
  }

  return { chunks: resultChunks, links: resultLinks };
}

/**
 * Восстанавливает текст из чанков с сохранением ссылок
 * @param chunks Массив чанков
 * @param links Массив ссылок
 * @returns Восстановленный текст
 */
export function restoreTextWithLinks(chunks: string[], links: string[]): string {
  let result = '';
  for (let i = 0; i < chunks.length; i++) {
    result += chunks[i];
    if (i < links.length) {
      result += links[i];
    }
  }
  return result;
}

/**
 * Нормализует текст для перевода
 * @param text Текст для нормализации
 * @returns Нормализованный текст
 */
export function normalizeText(text: string): string {
  return text
    .replace(/\r\n/g, '\n') //统一换行符
    .replace(/\t/g, '    ') // Заменяем табы на пробелы
    .replace(/  +/g, ' '); // Убираем множественные пробелы
}

/**
 * Проверка валидности текста для перевода
 * @param text Текст для проверки
 * @returns true если текст валиден
 */
export function isValidText(text: string): boolean {
  return typeof text === 'string' && text.length > 0;
}

File: translation-library/python/src/translation_library/chunker.py
"""
Модуль для разбивки текста на чанки
Основан на реализации из SillyTavern
"""

import re
from typing import List, Tuple


def chunk_text(text: str, max_size: int) -> List[str]:
    """
    Разбивает текст на чанки заданного размера
    
    Args:
        text: Текст для разбивки
        max_size: Максимальный размер чанка в символах
        
    Returns:
        Массив чанков
    """
    if not text or len(text) <= max_size:
        return [text]
    
    chunks: List[str] = []
    position = 0
    
    while position < len(text):
        # Пытаемся разбить по предложению или абзацу
        chunk_end = min(position + max_size, len(text))
        
        # Если это не конец текста, пытаемся найти границу предложения
        if chunk_end < len(text):
            # Ищем точку, восклицательный или вопросительный знак
            sentence_end = re.search(r'[.!?]\s', text[position:chunk_end])
            if sentence_end and sentence_end.start() > max_size * 0.5:
                chunk_end = position + sentence_end.end()
            else:
                # Ищем перенос строки
                newline_end = re.search(r'\n\s*', text[position:chunk_end])
                if newline_end and newline_end.start() > max_size * 0.5:
                    chunk_end = position + newline_end.start()
        
        chunk = text[position:chunk_end].strip()
        if chunk:
            chunks.append(chunk)
        
        position = chunk_end
    
    return chunks if chunks else [text]


def split_by_sentences(text: str) -> List[str]:
    """
    Разбивает текст по предложениям
    
    Args:
        text: Текст для разбивки
        
    Returns:
        Массив предложений
    """
    # Регулярное выражение для разделения по предложениям
    sentence_regex = r'([^.!?]+[.!?]+["\']?|\S+)'
    matches = re.findall(sentence_regex, text)
    
    return [s.strip() for s in matches if s.strip()]


def merge_chunks(chunks: List[str]) -> str:
    """
    Объединяет чанки обратно в текст
    
    Args:
        chunks: Массив чанков
        
    Returns:
        Объединённый текст
    """
    return ' '.join(chunks)


def chunk_with_links(text: str, max_size: int) -> Tuple[List[str], List[str]]:
    """
    Сохраняет ссылки на изображения при чанковании
    
    Args:
        text: Текст для чанкования
        max_size: Максимальный размер чанка
        
    Returns:
        Объект с чанками и ссылками
    """
    # Разделяем текст по ссылкам на изображения
    image_regex = r'!\[.*?\]\([^)]*\)'
    chunks = re.split(image_regex, text)
    links = re.findall(image_regex, text)
    
    result_chunks: List[str] = []
    result_links: List[str] = []
    
    for i, chunk in enumerate(chunks):
        if not chunk or not chunk.strip():
            continue
        
        chunk_chunks = chunk_text(chunk, max_size)
        result_chunks.extend(chunk_chunks)
        
        if i < len(links):
            result_links.append(links[i])
    
    return result_chunks, result_links


def restore_text_with_links(chunks: List[str], links: List[str]) -> str:
    """
    Восстанавливает текст из чанков с сохранением ссылок
    
    Args:
        chunks: Массив чанков
        links: Массив ссылок
        
    Returns:
        Восстановленный текст
    """
    result = ''
    for i, chunk in enumerate(chunks):
        result += chunk
        if i < len(links):
            result += links[i]
    return result


def normalize_text(text: str) -> str:
    """
    Нормализует текст для перевода
    
    Args:
        text: Текст для нормализации
        
    Returns:
        Нормализованный текст
    """
    return (text
            .replace('\r\n', '\n')  # Унифицируем переносы строк
            .replace('\t', '    ')   # Заменяем табы на пробелы
            .replace('  +', ' '))    # Убираем множественные пробелы


def is_valid_text(text: str) -> bool:
    """
    Проверка валидности текста для перевода
    
    Args:
        text: Текст для проверки
        
    Returns:
        True если текст валиден
    """
    return isinstance(text, str) and len(text) > 0